server/sonar-server/src/main/java/org/sonar/server/usertoken/ws/CreateAction.java

```diff
@@ -0,0 +1,142 @@
+import org.sonar.server.token.TokenGenerator;
+import org.sonar.server.user.UserSession;
+import org.sonar.server.user.index.UserIndex;
+import org.sonarqube.ws.WsUserTokens.CreateWsResponse;
+import org.sonarqube.ws.client.usertoken.CreateWsRequest;
+
+import static java.lang.String.format;
+import static org.sonar.api.utils.DateUtils.formatDateTime;
+import static org.sonar.server.ws.WsUtils.writeProtobuf;
+import static org.sonarqube.ws.client.usertoken.UserTokensWsParameters.PARAM_LOGIN;
+import static org.sonarqube.ws.client.usertoken.UserTokensWsParameters.PARAM_NAME;
+
+public class CreateAction implements UserTokensWsAction {
+  private final DbClient dbClient;
+  private final UserSession userSession;
+  private final System2 system;
+  private final TokenGenerator tokenGenerator;
+  private final UserIndex userIndex;
+
+  public CreateAction(UserSession userSession, DbClient dbClient, System2 system, TokenGenerator tokenGenerator, UserIndex userIndex) {
+    this.userSession = userSession;
+    this.dbClient = dbClient;
+    this.system = system;
+    this.tokenGenerator = tokenGenerator;
+    this.userIndex = userIndex;
+  }
+
+  @Override
+  public void define(WebService.NewController context) {
+    WebService.NewAction action = context.createAction("create")
+      .setSince("5.3")
+      .setDescription("Create a new random token")
+      .setResponseExample(getClass().getResource("create-example.json"))
+      .setHandler(this);
+
+    action.createParam("login")
+      .setRequired(true)
+      .setDescription("User login")
+      .setExampleValue("g.hopper");
+
+    action.createParam("name")
+      .setRequired(true)
+      .setDescription("Token description")
+      .setExampleValue("SonarQube analysis");
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    CreateWsResponse createWsResponse = doHandle(toCreateWsRequest(request));
+    writeProtobuf(createWsResponse, request, response);
+  }
+
+  private CreateWsResponse doHandle(CreateWsRequest request) {
+    userSession.checkGlobalPermission(GlobalPermissions.SYSTEM_ADMIN);
+
+    DbSession dbSession = dbClient.openSession(false);
+    try {
+      String token = tokenGenerator.generateToken();
+      String hashedToken = tokenGenerator.hash(token);
+
+      checkRequest(dbSession, request, hashedToken);
+
+      UserTokenDto userTokenDto = buildUserTokenDto(request, hashedToken);
+      insertTokenInDb(dbSession, userTokenDto);
+
+      return buildResponse(userTokenDto, token);
+    } finally {
+      dbClient.closeSession(dbSession);
+    }
+  }
+
+  private void checkRequest(DbSession dbSession, CreateWsRequest request, String hashedToken) {
+    // checks login exists
+    userIndex.getByLogin(request.getLogin());
+
+    Optional<UserTokenDto> userTokenDto = dbClient.userTokenDao().selectByLoginAndNameOrHashedToken(dbSession, request.getLogin(), request.getName(), hashedToken);
```

Comment:
I wanted to avoid to do 2 database queries, that's why I combined them. The error message is not good enought though. Anyway, I separate in 2 db queries